### PR TITLE
Properly populate transaction fees

### PR DESCRIPTION
**Pull request by clarete**
# TL;DR
Migration to fill in all the fee columns so the expression `netAmount = amount + hostFee + platformFee + paymentProcessorFee` is true for every single transaction in the database.

## Details
A single real transaction is currently represented by two rows in the `Transactions` table. One of the type `DEBIT` to represent the change in the **User ledger** and another one of the type `CREDIT` to represent changes in the **Collective Ledger**.

That's all good so far. The representation of the fees, however, is a little more confusing. Both rows (`CREDIT` & `DEBIT`) contain the columns to store the fees. But they're currently only populated on the `CREDIT` row. And their values are positive, so they don't really express exactly what is going on with that transaction. To make it a little easier to understand. Here's how a transaction is currently stored:

### How one transaction is stored

|            | User Ledger | Collective Ledger |
|------------|------------:|------------------:|
| Type       |       DEBIT |            CREDIT |
| Amount     |       -4075 |              5000 |
| Host Fee   |           0 |               500 |
| Plat Fee   |           0 |               250 |
| PP Fee     |           0 |               175 |
| Net Amount |       -5000 |              4075 |

### How it should look after the migration

|            | User Ledger | Collective Ledger |
|------------|------------:|------------------:|
| Type       |       DEBIT |            CREDIT |
| Amount     |       -4075 |              5000 |
| Host Fee   |        -500 |              -500 |
| Plat Fee   |        -250 |              -250 |
| PP Fee     |        -175 |              -175 |
| Net Amount |       -5000 |              4075 |

## Motivation

There are two features that we want to add that will greatly benefit from this change:

1. Refund transactions. There's an ongoing effort to automate refunding transactions. Although the feature is [already implemented and ready to be deployed](https://github.com/opencollective/opencollective-api/pull/1134), it greatly increases the complexity of managing transactions and fees. Especially when Payment Processing Fees are not refunded.

2. We're planning to unfold each transaction between User & Collective into more transactions between *User & Payment Processor*, *Payment Processor & Host* and *Host & Collective*. Having the fees correctly filled in will make this job a few orders of magnitude easier

## Challenges

1. **Currency conversion**. The formula `netAmount = amount + hostFee + platformFee + paymentProcessorFee` doesn't apply for transactions between hosts and collectives that use different currencies. That's OK for 93% of our transactions because they're between USD & USD but we need to improve the formula to account for the conversion.

2. **Manual changes**. Some database entries have the fees registered in both types of transactions (`DEBIT` and `CREDIT`) so some care has to be taken to make sure that we don't automate the replication of mistakes made by manual changes.

### CODE REVIEW

**Comment by asood123** [comment]
@clarete let me know when you want to talk through currency mismatch cases... I can help you whittle them down.

**Comment by clarete** [comment]
Just documenting what's going on with the off by one errors we found on the ledger. What I initially thought that was caused by manual mistakes ended up being a little deeper. The patterns that lead me to find it were:

* `netAmountICC` off by either `1` or `-1`
* all the transactions were of type `CREDIT`

That made me check the way we calculated the fees and I found out that we don't really round the fees. And then when we store the value we lose the precision because we're storing the currency related values as cents. One example that I have is the following one (all currencies involved are in `USD`):

### Example Input
* `amount 21650`
* `platform fee 0.05`
* `host fee 0.05`

### What `createFromPayload`/`createDoubleEntry` are currently returning:

#### CREDIT
* `amountIHC 21650`
* `netAmountICC 19485 `
#### DEBIT
* `amountIHC -19485`
* `netAmountICC -21650`

### What it should return instead

#### CREDIT
* `amountIHC 21650`
* `netAmountICC 19486 `

#### DEBIT
* `amountIHC -21650`
* `netAmountICC -19486 `

**Comment by asood123** [comment]
Ah good catch!!

File: scripts/properly_populate_transaction_fees.js [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+ * transactions at once is the fastest way to run it and get the
+ * output, however it also exhausts node's memory limits with batches
+ * larger than ~50000. Use --max-old-space=8192 to do it all in one
+ * batch and iterate faster.
+ *
+ * For production usage, It might be a good idea to break it down in
+ * multiple batches to leave some room for the database to process
+ * operations from the web clients. I've been running 7 batches of
+ * 10000 and didn't take more than ~20s in my machine (i7~2.2GHz),
+ * which is just a little bit slower than running it all in one batch.
```
Comment (asood123) [inline scripts/properly_populate_transaction_fees.js]: Nice to have these perf numbers!

File: scripts/properly_populate_transaction_fees.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    return transactions;
+  }
+
+  /** Convert `value` to negative if it's possitive */
+  toNegative = (value) => value > 0 ? -value : value;
+
+  /** Ensure that `tr` has the `hostCurrencyFxRate` field filled in */
+  ensureHostCurrencyFxRate = (tr) => {
+    if (tr.amount === tr.amountInHostCurrency && !tr.hostCurrencyFxRate)
+      tr.hostCurrencyFxRate = 1;
```
Comment (asood123) [inline scripts/properly_populate_transaction_fees.js]: This is a minor thing but I feel like, in general, it's better to make these type of functions as `get` functions, instead of modifying the underlying model. If it's `getHostCurrencyFXRate`, you don't need to think about whether model was saved or not. Otherwise, whoever calls this function will need to remember that it modifies the passed in object and they need to save it.

File: scripts/properly_populate_transaction_fees.js [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  hasHostFee = (tr) => tr.hostFeeInHostCurrency
+    && parseInt(tr.hostFeeInHostCurrency, 10) !== 0;
+  /** Return true if the transaction has any platform fees */
+  hasPlatformFee = (tr) => tr.paymentProcessorFeeInHostCurrency
+    && parseInt(tr.paymentProcessorFeeInHostCurrency, 10) !== 0;
+  /** Return false if there are no fees on a transaction */
+  hasFees = (tr) => this.hasHostFee(tr) || this.hasPlatformFee(tr)
+    || (tr.platformFeeInHostCurrency && parseInt(tr.platformFeeInHostCurrency, 10) !== 0);
+
+  rewriteFees = (credit, debit) => {
```
Comment (asood123) [inline scripts/properly_populate_transaction_fees.js]: Same comment here. I think `getFees` would be cleaner. It returns the fees and the caller is clearly responsible for setting and storing them. 

Since this is just a script, it's fine to leave it as it is. More a comment for future reusable components.

File: scripts/properly_populate_transaction_fees.js [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        console.log('Expense.: true, true');
+        return;
+      }
+
+      // this.rewriteFees(credit, debit);
+
+      console.log('  Expense.:', transactionsLib.verify(tr1), transactionsLib.verify(tr2));
+
+      if (!transactionsLib.verify(credit)) {
+        console.log(`| EDAU | CREDIT | ${credit.id} | ${credit.TransactionGroup} | ${transactionsLib.difference(credit)} |`);
```
Comment (asood123) [inline scripts/properly_populate_transaction_fees.js]: What's EDAU?

File: server/paymentProviders/stripe/bitcoin.js [hunk 5]
```diff
@@ -84,8 +87,8 @@ export default {
             hostCurrency: balanceTransaction.currency,
             amountInHostCurrency: balanceTransaction.amount,
             hostCurrencyFxRate: order.totalAmount / balanceTransaction.amount,
-            hostFeeInHostCurrency: parseInt(balanceTransaction.amount * hostFeePercent / 100, 10),
-            platformFeeInHostCurrency: order.totalAmount*0.05,
+            hostFeeInHostCurrency,
+            platformFeeInHostCurrency: Math.round(order.totalAmount * 0.05),  /* Why not dividing by 100? Is it in a different unit? */
```
Comment (asood123) [inline server/paymentProviders/stripe/bitcoin.js]: That's a good question on why it's not divided by 100. Can you look at a couple of bitcoin transactions in the database and see if fees look reasonable?

File: scripts/properly_populate_transaction_fees.js (hunk 2 above)
Comment (clarete) [reply scripts/properly_populate_transaction_fees.js]: Yeah, in that case what I usually like is to create two layers.

 1. Function for the calculation of the new value (the `get` function you mentioned);
 2. Function that actually performs the mutation in the field (the one that I did);

I started from the second layer because I really wanted to test the actual side effect since these objects are going to be instances of database rows and whatever change to them will actually reflect in the database.

Although this is a script, I actually ended up testing most of the delicate functions because I was thinking about keeping that script around since it became sort of a cleanup tool. But I totally agree with you that we should also have the `get*` functions within a library (thinking of `lib/transactions`). I'll do that in small chunks for things that I need to reuse during the changes in the application code.

File: scripts/properly_populate_transaction_fees.js (hunk 3 above)
Comment (clarete) [reply scripts/properly_populate_transaction_fees.js]: I wrote an answer to the previous question that might address this too.

File: scripts/properly_populate_transaction_fees.js (hunk 4 above)
Comment (clarete) [reply scripts/properly_populate_transaction_fees.js]: EDAU: Expense Don't Add Up
ODAU: Order Don't Add Up

File: scripts/properly_populate_transaction_fees.js (hunk 1 above)
Comment (clarete) [reply scripts/properly_populate_transaction_fees.js]: I started measuring because it exhausted the v8 memory when I joined the Collectives table to read the `hostFeePercent` but I think it also helps knowing what to expect when we run it on prod.

File: server/paymentProviders/stripe/bitcoin.js (hunk 5 above)
Comment (clarete) [reply server/paymentProviders/stripe/bitcoin.js]: Good call. I'll check that and I'll also double check if there's any other place that I should be rounding it up.

**Comment by clarete** [comment]
The failure in the tests is related to the decrease in the test coverage. I'll come back to it as soon as the migration is rolled out.
